src/batou_ext/nixos.py

```diff
@@ -1,86 +1,75 @@
 
 from batou.component import Component
-from batou.lib.file import Purge
+from batou.lib.file import JSONContent, ManagedContentBase, Purge
 
-from batou_ext.nix import NixFile, component_to_nix, nix_dict_to_nix
+import batou_ext.nix
 
-# XXX: error messages stemming from the "batouModule" are displayed with
-# wrong location information, it shows the glue module instead of the actual
-# module. Can we fix that somehow?
-GLUE_MODULE_TEMPLATE = """\
-{{ pkgs, ... }}@args:
 
-with builtins;
+class NixosModuleContent(ManagedContentBase):
+    json_file: Optional[Component] = None
 
-let
-  moduleArgs = (import {workdir}/{prefix}_generated_context.nix) // args;
-  batouModule = import {workdir}/{name}.nix moduleArgs;
-in
-{{ imports = [ batouModule ]; }}
-"""
+    def render(self):
+        self.content = self.expand(
+            """\
+# this file is managed by batou, any manual changes will be overwritten by the next deployment
+{pkgs, ...}@args: let
+  component = builtins.fromJSON (builtins.readFile {{component.json_file.path}});
+  module = (
 
+{{component.content}}
 
-class NixOSModuleContext(Component):
-
-    source_component: Component = None
-    prefix: str = None
-
-    def configure(self):
-
-        if self.source_component:
-            component = self.source_component
-        else:
-            component = self.parent
-
-        context = nix_dict_to_nix({"component": component_to_nix(component)})
-
-        if self.prefix is None:
-            self.prefix = component.__class__.__name__.lower()
-
-        self += NixFile(
-            f"{self.prefix}_generated_context.nix",
-            content=context,
-            format_nix_code=True,
-            sensitive_data=True,
+  );
+in {
+  imports = [(module (args // { inherit component; }))];
+}
+            """,
         )
 
 
-class NixOSModule(Component):
-
-    namevar = "name"
-
-    name: str
-    path: Path = Path("/etc/local/nixos")
-    context = None
+@batou_ext.nix.rebuild
+class NixosModule(Component):
+    namevar = "path"
+    source_component: Optional[Component] = None
 
     def configure(self):
-
-        self += NixFile(f"{self.name}.nix")
-
-        if self.context is None:
-            if hasattr(self.parent, "nixos_context"):
-                self.context = self.parent.nixos_context
-            else:
-                self.context = NixOSModuleContext(source_component=self.parent)
-                self.parent.nixos_context = self.context
-                self.parent += self.context
-
-        self += NixFile(
-            self.path / f"batou_{self.name}.nix",
-            content=GLUE_MODULE_TEMPLATE.format(
-                workdir=self.workdir, name=self.name, prefix=self.context.prefix
+        basename = os.path.basename(self.path)
+
+        self.json_file = JSONContent(
+            self.map(basename.replace(".nix", "_component.json")),
+            data=json.loads(
+                json.dumps(
+                    self.source_component or self.parent,
```

Comment:
I think this needs special casing for `Address`/`NetLoc`.

Reply:
A lot of components will require special handling, especially custom ones. I'm very much not sure how to handle this properly. The question is where do you draw the line honestly. I'd prefer as little special handling as possible.

Reply:
>  I'd prefer as little special handling as possible.

I don't disagree.
But it seems as if Address isn't evne JSON serializable, so currently the NixOSModule code would fail with such components, correct?
Ideally one could make classes JSON serializable in Python, but I'd have to do some more research, i.e. not now.